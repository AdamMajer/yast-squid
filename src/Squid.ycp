/* ------------------------------------------------------------------------------
 * Copyright (c) 2006 Novell, Inc. All Rights Reserved.
 *
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of version 2 of the GNU General Public License as published by the
 * Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, contact Novell, Inc.
 *
 * To contact Novell about this file by physical or electronic mail, you may find
 * current contact information at www.novell.com.
 * ------------------------------------------------------------------------------
 */

/**
 * File:	modules/Squid.ycp
 * Package:	Configuration of squid
 * Summary:	Squid settings, input and output functions
 * Authors:	Daniel Fiser <dfiser@suse.cz>
 *
 * $Id: Squid.ycp 27914 2006-02-13 14:32:08Z locilka $
 *
 * Representation of the configuration of squid.
 * Input and output routines.
 */

{

module "Squid";
textdomain "squid";

import "Progress";
import "Report";
import "Summary";
import "Message";

path squid_path = .squid;


/**
 * Data was modified?
 */
boolean modified = false;

/**
 * Map of all configuration settings except consequential
 */
map<string,any> settings = $[];

/**
 * List of http_ports. Format:
 * [ $["host" : "hostname",
 *     "port" : 3128,
 *     "transparent" : true],
 *     ...
 * ]
 */
list <map <string,any> > http_ports = [];

/** TODO
 * List of acls. Format:
 * [ $[ ???
 */
list <map <string,any> > acls = [];

/**
 * List of access control parameters. Format:
 * [ $["allow" : true,
 *     "acl" : ["acl1", "!acl2", ...] ],
 *     ...
 * ]
 */
list <map <string,any> > http_accesses = [];

/**
 * List of refresh patterns. Format:
 * [ $["regexp" : "^ftp:",
 *     "min" : 12,
 *     "max" : 12,
 *     "percent" : 12],
 *     ...
 * ]
 */
list <map <string,any> > refresh_patterns = [];




global boolean Modified(){
    return modified;
}


/**
 */
global boolean proposal_valid = false;

/**
 * Write only, used during autoinstallation.
 * Don't run services and SuSEconfig, it's all done at one place.
 */
global boolean write_only = false;

/**
 * Abort function
 * return boolean return true if abort
 */
global boolean() AbortFunction = Modified;

/**
 * Abort function
 * @return boolean return true if abort
 */
global define boolean Abort() ``{
    if(AbortFunction != nil)
    {
	return AbortFunction () == true;
    }
    return false;
}


/**
 * Read setting of parameter http_port. It can be in form:
 * http_port [hostname:]port [transparent]
 *
 * return true on success
 */
boolean readHttpPorts()
{
    list tmp = [];
    list tmp2 = [];
    map <string,any> tmp_http_port = $[];

    http_ports = [];
    foreach (string value, (list<string>)SCR::Read(add(squid_path,"http_port")),
        {
            tmp_http_port = $[];
            tmp = [];
            tmp2 = [];

            tmp = splitstring(value," \t");

            //can parse only 'http_port hostname:port [transparent]'
            if (size(tmp) < 1 || size(tmp) > 2) return false;

            // hostname and port
            tmp2 = splitstring(tmp[0]:"",":");
            y2debug("readHttpPorts - tmp2: %1",tmp2);
            if (size(tmp2) == 1){
                tmp_http_port["host"] = "";
                tmp_http_port["port"] = tointeger(tmp2[0]:"");
            }else{
                tmp_http_port["host"] = (string)tmp2[0]:"";
                tmp_http_port["port"] = tointeger(tmp2[1]:"");
            }

            //transparent option
            if (size(tmp) == 2 && tmp[1]:"" == "transparent")
                tmp_http_port["transparent"] = true;

            http_ports = add(http_ports, tmp_http_port);
        });

    return true;
}

/**
 * Read all squid settings
 * @return true on success
 */
global boolean Read() {
    boolean ok = true;

    /* Squid read dialog caption */
    string caption = _("Initializing squid Configuration");

    integer steps = 2;

    Progress::New(caption, " ", steps, [_("Read the settings")], [_("Reading the settings..."), _("Finished")], "");

    if(Abort()) return false;
    Progress::NextStage();

    if (!readHttpPorts()) ok = false;
    // TODO:
    //if (!readRefreshPatterns()) ok = false;
    //if (!readACLs()) ok = false;
    //if (!readHttpAccesses()) ok = false;
    //if (!readRestSetting()) ok = false;

    y2debug("Squid::Read - http_ports: %1",http_ports);

    /* Error message */
    if(!ok) Report::Error(_("Cannot read configuration file."));

    Progress::NextStep();

    return true;
}

/**
 * Write all squid settings
 * @return true on success
 */
global boolean Write() {

    /* Squid read dialog caption */
    string caption = _("Saving squid Configuration");

    // TODO FIXME And set the right number of stages
    integer steps = 2;

    integer sl = 500;
    sleep(sl);

    // TODO FIXME Names of real stages
    // We do not set help text here, because it was set outside
    Progress::New(caption, " ", steps, [
	    /* Progress stage 1/2 */
	    _("Write the settings"),
	    /* Progress stage 2/2 */
	    _("Run SuSEconfig")
	], [
	    /* Progress step 1/2 */
	    _("Writing the settings..."),
	    /* Progress step 2/2 */
	    _("Running SuSEconfig..."),
	    /* Progress finished */
	    _("Finished")
	],
	""
    );

    // write settings
    if(Abort()) return false;
    Progress::NextStage();
    /* Error message */
    if(false) Report::Error (_("Cannot write settings."));
    sleep(sl);

    // run SuSEconfig
    if(Abort()) return false;
    Progress::NextStage ();
    /* Error message */
    if(false) Report::Error (Message::SuSEConfigFailed());
    sleep(sl);

    if(Abort()) return false;
    /* Progress finished */
    Progress::NextStage();
    sleep(sl);

    if(Abort()) return false;
    return true;
}

/**
 * Get all squid settings from the first parameter
 * (For use by autoinstallation.)
 * @param settings The YCP structure to be imported.
 * @return boolean True on success
 */
global boolean Import (map settings) {
    // TODO FIXME: your code here (fill the above mentioned variables)...
    return true;
}

/**
 * Dump the squid settings to a single map
 * (For use by autoinstallation.)
 * @return map Dumped settings (later acceptable by Import ())
 */
global map Export () {
    // TODO FIXME: your code here (return the above mentioned variables)...
    return $[];
}

/**
 * Create a textual summary and a list of unconfigured cards
 * @return summary of the current configuration
 */
global list Summary() {
    // TODO FIXME: your code here...
    /* Configuration summary text for autoyast */
    return [ _("Configuration summary..."), [] ];
}

/**
 * Create an overview table with all configured cards
 * @return table items
 */
global list Overview() {
    // TODO FIXME: your code here...
    return [];
}

/**
 * Return packages needed to be installed and removed during
 * Autoinstallation to insure module has all needed software
 * installed.
 * @return map with 2 lists.
 */
global map AutoPackages() {
    // TODO FIXME: your code here...
    return $[ "install":[], "remove":[] ];
}

/* EOF */
}
/* vim: set sw=4 ts=4 et ft=ycp : */
