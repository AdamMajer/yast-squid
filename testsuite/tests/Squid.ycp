{

// testedfiles: Squid.ycp

include "testsuite.ycp";
TESTSUITE_INIT([], nil);

import "Squid";

DUMP("Squid::Modified");
TEST(``(Squid::Modified()), [], nil);

}
