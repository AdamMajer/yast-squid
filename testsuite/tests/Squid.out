Read	.etc.squid."http_port" [["localhost:3128"], ["80", "transparent"]]
Read	.etc.squid."refresh_pattern" [["^ftp:", "1440", "20%", "10080"], ["-i", "^gopher:", "1440", "0%", "1440"], [".", "0", "20%", "4320"]]
Read	.etc.squid."acl" [["QUERY", "urlpath_regex", "cgi-bin", "\\?"], ["apache", "rep_header", "Server", "^Apache"], ["all", "src", "0.0.0.0/0.0.0.0"], ["manager", "proto", "cache_object"], ["localhost", "src", "127.0.0.1/255.255.255.255"], ["localhost_public", "src", "10.20.1.241/255.255.255.255"], ["to_localhost", "dst", "127.0.0.0/8"], ["SSL_ports", "port", " 443"], ["Safe_ports", "port", "80"], ["Safe_ports", "port", "21"], ["Safe_ports", "port", "443"], ["Safe_ports", "port", "70"], ["Safe_ports", "port", "210"], ["Safe_ports", "port", "1025-65535"], ["Safe_ports", "port", "280"], ["Safe_ports", "port", "488"], ["Safe_ports", "port", "591"], ["Safe_ports", "port", "777"], ["CONNECT", "method", "CONNECT"]]
Read	.etc.squid."http_access" [["allow", "manager", "localhost"], ["deny", "manager"], ["deny", "!Safe_ports"], ["deny", "CONNECT", "!SSL_ports"], ["allow", "localhost"], ["allow", "localhost_public"], ["deny", "all"]]
Read	.etc.squid."access_log" [["/var/log/squid/access.log"]]
Read	.etc.squid."cache_dir" [["ufs", "/var/cache/squid", "100", "16", "256"]]
Read	.etc.squid."cache_log" [["/var/log/squid/cache.log"]]
Read	.etc.squid."cache_mem" [["80", "MB"]]
Read	.etc.squid."cache_mgr" [["webmaster"]]
Read	.etc.squid."cache_replacement_policy" [["lru"]]
Read	.etc.squid."cache_store_log" [["/var/log/squid/store.log"]]
Read	.etc.squid."cache_swap_high" [["95"]]
Read	.etc.squid."cache_swap_low" [["90"]]
Read	.etc.squid."client_lifetime" [["1", "day"]]
Read	.etc.squid."connect_timeout" [["2", "minutes"]]
Read	.etc.squid."emulate_httpd_log" [["off"]]
Read	.etc.squid."error_directory" [["/usr/share/squid/errors/English"]]
Read	.etc.squid."ftp_passive" [["on"]]
Read	.etc.squid."maximum_object_size" [["4096", "KB"]]
Read	.etc.squid."memory_replacement_policy" [["heap", "GDSF"]]
Read	.etc.squid."minimum_object_size" [["0", "KB"]]
Dump	==================================================
Dump	=====================  ACL  ======================
Dump	==================================================
Dump	
Dump	GetACLs()
Return	[$["name":"QUERY", "options":["cgi-bin \\?"], "type":"urlpath_regex"], $["name":"apache", "options":["Server", "^Apache"], "type":"rep_header"], $["name":"all", "options":["0.0.0.0/0.0.0.0"], "type":"src"], $["name":"manager", "options":["cache_object"], "type":"proto"], $["name":"localhost", "options":["127.0.0.1/255.255.255.255"], "type":"src"], $["name":"localhost_public", "options":["10.20.1.241/255.255.255.255"], "type":"src"], $["name":"to_localhost", "options":["127.0.0.0/8"], "type":"dst"], $["name":"SSL_ports", "options":[" 443"], "type":"port"], $["name":"Safe_ports", "options":["80"], "type":"port"], $["name":"Safe_ports", "options":["21"], "type":"port"], $["name":"Safe_ports", "options":["443"], "type":"port"], $["name":"Safe_ports", "options":["70"], "type":"port"], $["name":"Safe_ports", "options":["210"], "type":"port"], $["name":"Safe_ports", "options":["1025-65535"], "type":"port"], $["name":"Safe_ports", "options":["280"], "type":"port"], $["name":"Safe_ports", "options":["488"], "type":"port"], $["name":"Safe_ports", "options":["591"], "type":"port"], $["name":"Safe_ports", "options":["777"], "type":"port"], $["name":"CONNECT", "options":["CONNECT"], "type":"method"]]
Dump	
Dump	GetACL(0)
Return	$["name":"QUERY", "options":["cgi-bin \\?"], "type":"urlpath_regex"]
Dump	GetACL(2)
Return	$["name":"all", "options":["0.0.0.0/0.0.0.0"], "type":"src"]
Dump	GetACL(100) - out of range
Return	$[]
Dump	++++++++++++++++++++++++++++++++++++++++++++++++++
Dump	
Dump	AddACL("name", "type", ["list", "of", "options"])
Return	nil
Return	$["name":"name", "options":["list", "of", "options"], "type":"type"]
Dump	
Dump	ModifyACL(0, "QUERY2", "urlpath_regex", ["cgi-bin \? aaa"])
Return	$["name":"QUERY", "options":["cgi-bin \\?"], "type":"urlpath_regex"]
Return	nil
Return	$["name":"QUERY2", "options":["cgi-bin \\? aaa"], "type":"urlpath_regex"]
Return	nil
Dump	
Dump	ModifyACL(100, "A", "a", ["a"]) - out of range
Return	$[]
Return	nil
Return	$[]
Dump	
Dump	DelACL(1)
Return	$["name":"apache", "options":["Server", "^Apache"], "type":"rep_header"]
Return	$["name":"all", "options":["0.0.0.0/0.0.0.0"], "type":"src"]
Return	nil
Return	$["name":"all", "options":["0.0.0.0/0.0.0.0"], "type":"src"]
Dump	
Dump	DelACL(100) - out of range
Return	nil
Return	$[]
Return	$["name":"QUERY", "options":["cgi-bin \\?"], "type":"urlpath_regex"]
Dump	++++++++++++++++++++++++++++++++++++++++++++++++++
Dump	
Dump	NumACLs(0)
Return	1
Dump	NumACLs(10)
Return	10
Dump	
Dump	ACLIsUsedBy(0) - QUERY
Dir	.etc.squid: ["access_log", "acl", "cache_dir", "cache_log", "cache_mem", "cache_mgr", "cache_replacement_policy", "cache_store_log", "cache_swap_high", "cache_swap_log", "cache_swap_low", "client_lifetime", "connect_timeout", "emulate_httpd_log", "error_directory", "ftp_passive", "http_access", "http_port", "maximum_object_size", "memory_replacement_policy", "minimum_object_size", "no_cache", "refresh_pattern"]
Read	.etc.squid."no_cache" [["deny", "QUERY"]]
Read	.etc.squid."access_log" [["/var/log/squid/access.log"]]
Return	["no_cache"]
Dump	ACLIsUsedBy(1)
Dir	.etc.squid: ["access_log", "acl", "cache_dir", "cache_log", "cache_mem", "cache_mgr", "cache_replacement_policy", "cache_store_log", "cache_swap_high", "cache_swap_log", "cache_swap_low", "client_lifetime", "connect_timeout", "emulate_httpd_log", "error_directory", "ftp_passive", "http_access", "http_port", "maximum_object_size", "memory_replacement_policy", "minimum_object_size", "no_cache", "refresh_pattern"]
Read	.etc.squid."no_cache" [["deny", "QUERY"]]
Read	.etc.squid."access_log" [["/var/log/squid/access.log"]]
Return	["http_access"]
Dump	==================================================
Dump	================  HTTP_ACCESS  ===================
Dump	==================================================
Dump	
Dump	GetHttpAccesses()
Return	[$["acl":["manager", "localhost"], "allow":true], $["acl":["manager"], "allow":false], $["acl":["!Safe_ports"], "allow":false], $["acl":["CONNECT", "!SSL_ports"], "allow":false], $["acl":["localhost"], "allow":true], $["acl":["localhost_public"], "allow":true], $["acl":["all"], "allow":false]]
Dump	
Dump	GetHttpAccess(0)
Return	$["acl":["manager", "localhost"], "allow":true]
Dump	GetHttpAccess(2)
Return	$["acl":["!Safe_ports"], "allow":false]
Dump	GetHttpAccess(100) - out of range
Return	$[]
Dump	++++++++++++++++++++++++++++++++++++++++++++++++++
Dump	
Dump	AddHttpAccess(true, ["list", "of", "acls"])
Return	nil
Return	$["acl":["list", "of", "acls"], "allow":true]
Dump	
Dump	ModifyHttpAccess(0, false, ["manager", "locahost"])
Return	$["acl":["manager", "localhost"], "allow":true]
Return	nil
Return	$["acl":["manager", "localhost"], "allow":false]
Return	nil
Dump	
Dump	ModifyHttpAccess(100, true, ["a"]) - out of range
Return	$[]
Return	nil
Return	$[]
Dump	
Dump	DelHttpAccess(1)
Return	$["acl":["manager"], "allow":false]
Return	$["acl":["!Safe_ports"], "allow":false]
Return	nil
Return	$["acl":["!Safe_ports"], "allow":false]
Dump	
Dump	MoveHttpAccess(0,1)
Return	$["acl":["manager", "localhost"], "allow":true]
Return	$["acl":["!Safe_ports"], "allow":false]
Return	nil
Return	$["acl":["!Safe_ports"], "allow":false]
Return	$["acl":["manager", "localhost"], "allow":true]
Dump	
Dump	DelHttpAccess(100) - out of range
Return	nil
Return	$[]
Return	$["acl":["!Safe_ports"], "allow":false]
Dump	==================================================
Dump	=================   SETTINGS  ====================
Dump	==================================================
Dump	
Dump	GetSettings()
Return	$["access_log":["/var/log/squid/access.log"], "cache_dir":["ufs", "/var/cache/squid", "100", "16", "256"], "cache_log":["/var/log/squid/cache.log"], "cache_mem":["80", "MB"], "cache_mgr":["webmaster"], "cache_replacement_policy":["lru"], "cache_store_log":["/var/log/squid/store.log"], "cache_swap_high":["95"], "cache_swap_low":["90"], "client_lifetime":["1", "days"], "connect_timeout":["2", "minutes"], "emulate_httpd_log":["off"], "error_directory":["/usr/share/squid/errors/English"], "ftp_passive":["on"], "maximum_object_size":["4096", "KB"], "memory_replacement_policy":["heap GDSF"], "minimum_object_size":["0", "KB"]]
Dump	
Dump	GetSetting("cache_dir")
Return	["ufs", "/var/cache/squid", "100", "16", "256"]
Dump	
Dump	SetSetting("cache_dir", ["uufs", "/var/", "10", "10", "10"])
Return	nil
Return	["uufs", "/var/", "10", "10", "10"]
Dump	==================================================
Dump	==============  REFRESH_PATTERNS  ================
Dump	==================================================
Dump	
Dump	GetRefreshPatterns()
Return	[$["case_sensitive":true, "max":"10080", "min":"1440", "percent":"20", "regexp":"^ftp:"], $["case_sensitive":false, "max":"1440", "min":"1440", "percent":"0", "regexp":"^gopher:"], $["case_sensitive":true, "max":"4320", "min":"0", "percent":"20", "regexp":"."]]
Dump	
Dump	GetRefreshPattern(0)
Return	$["case_sensitive":true, "max":"10080", "min":"1440", "percent":"20", "regexp":"^ftp:"]
Dump	GetRefreshPattern(2)
Return	$["case_sensitive":true, "max":"4320", "min":"0", "percent":"20", "regexp":"."]
Dump	GetRefreshPattern(100) - out of range
Return	$[]
Dump	++++++++++++++++++++++++++++++++++++++++++++++++++
Dump	
Dump	AddRefreshPattern("regexp", "100", "100", "100", false)
Return	nil
Return	$["case_sensitive":false, "max":"100", "min":"100", "percent":"100", "regexp":"regexp"]
Dump	
Dump	ModifyRefreshPattern(0, "regexp",  "100", "100", "100", false)
Return	$["case_sensitive":true, "max":"10080", "min":"1440", "percent":"20", "regexp":"^ftp:"]
Return	nil
Return	$["case_sensitive":false, "max":"100", "min":"100", "percent":"100", "regexp":"regexp"]
Dump	
Dump	ModifyRefreshPattern(100, "regexp",  "100", "100", "100", false)
Return	$[]
Return	nil
Return	$[]
Dump	
Dump	DelRefreshPattern(1)
Return	$["case_sensitive":false, "max":"1440", "min":"1440", "percent":"0", "regexp":"^gopher:"]
Return	$["case_sensitive":true, "max":"4320", "min":"0", "percent":"20", "regexp":"."]
Return	nil
Return	$["case_sensitive":true, "max":"4320", "min":"0", "percent":"20", "regexp":"."]
Dump	
Dump	MoveRefreshPattern(0,1)
Return	$["case_sensitive":false, "max":"100", "min":"100", "percent":"100", "regexp":"regexp"]
Return	$["case_sensitive":true, "max":"4320", "min":"0", "percent":"20", "regexp":"."]
Return	nil
Return	$["case_sensitive":true, "max":"4320", "min":"0", "percent":"20", "regexp":"."]
Return	$["case_sensitive":false, "max":"100", "min":"100", "percent":"100", "regexp":"regexp"]
Dump	
Dump	DelRefreshPattern(100) - out of range
Return	nil
Return	$[]
Return	$["case_sensitive":true, "max":"4320", "min":"0", "percent":"20", "regexp":"."]
Dump	==================================================
Dump	=================  HTTP_PORT  ====================
Dump	==================================================
Dump	
Dump	GetHttpPorts()
Return	[$["host":"localhost", "port":"3128"], $["host":"", "port":"80", "transparent":true]]
Dump	
Dump	GetHttpPort(0)
Return	$["host":"localhost", "port":"3128"]
Dump	GetHttpPort(2)
Return	$[]
Dump	GetHttpPort(100) - out of range
Return	$[]
Dump	++++++++++++++++++++++++++++++++++++++++++++++++++
Dump	
Dump	AddHttpPort("host", "port", true)
Return	nil
Return	$["host":"host", "port":"port", "transparent":true]
Dump	
Dump	ModifyHttpPort(0, "host", "port", true)
Return	$["host":"localhost", "port":"3128"]
Return	nil
Return	$["host":"host", "port":"port", "transparent":true]
Dump	
Dump	ModifyHttpPort(100, "host", "port", true)
Return	$[]
Return	nil
Return	$[]
Dump	
Dump	DelHttpPort(1)
Return	$["host":"", "port":"80", "transparent":true]
Return	$["host":"host", "port":"port", "transparent":true]
Return	nil
Return	$["host":"host", "port":"port", "transparent":true]
Dump	
Dump	DelHttpPort(100) - out of range
Return	nil
Return	$[]
Return	$["host":"host", "port":"port", "transparent":true]
